File: sdk/py/examples/ml/transformer.py
# OA_DOC_BEGIN: ml-transformer
import oa

vocabSize = 300
contextLength = 16
modelWidth = 32
hiddenWidth = 64
batchSize = 64
trainingSteps = 300
corpus = (
	"to be or not to be that is the question whether tis nobler in the mind "
	"to suffer the slings and arrows of outrageous fortune or to take arms "
	"against a sea of troubles and by opposing end them to be or not to be "
	"that is the question whether tis nobler in the mind to suffer the slings "
	"and arrows of outrageous fortune or to take arms against a sea of "
	"troubles and by opposing end them to be or not to be that is the question "
	"whether tis nobler in the mind to suffer the slings and arrows of "
	"outrageous fortune or to take arms against a sea of troubles and by "
	"opposing end them "
)

oa.FnMatrix.setRngSeed(20260714)
tokenizer = oa.BpeTokenizer(vocabSize)
tokenizer.train(corpus, vocabSize - 256)
assert tokenizer.vocabSize() == vocabSize
corpusTokens = tokenizer.encode(corpus)

model = oa.NnTransformer(
	vocabSize, contextLength, modelWidth, hiddenWidth,
)
parameters = model.allParameterPtrs()

def tokenMatrix(tokens, batch):
	return oa.FnMatrix.fromInt32(tokens, [batch, contextLength], oa.ScalarType.UInt32)

def nextBatch(tokens, cursor):
	limit = len(tokens) - contextLength - 1
	x, y = [], []
	for batch in range(batchSize):
		start = (cursor + batch * 7) % limit
		x.extend(tokens[start:start + contextLength])
		y.extend(tokens[start + 1:start + contextLength + 1])
	return tokenMatrix(x, batchSize), tokenMatrix(y, batchSize), (cursor + batchSize) % limit

def generate(tokenizer):
	prompt = "to be"
	context = (tokenizer.encode(prompt) + [0] * contextLength)[:contextLength]
	filled = max(1, min(len(tokenizer.encode(prompt)), contextLength))
	result = prompt
	for _ in range(32):
		logits = model.forward(tokenMatrix(context, 1))
		row = oa.FnMatrix.slice(logits, 0, filled - 1, filled)
		nextToken = int(oa.FnMatrix.argmax(row.reshape([vocabSize])))
		result += tokenizer.decode([nextToken])
		if filled < contextLength:
			context[filled] = nextToken
			filled += 1
		else:
			context = context[1:] + [nextToken]
	return result

optimizer = oa.AdamW(parameters, 0.01)
lossMetric = oa.MetricLoss()
progress = oa.CbProgressBar()
summary = oa.CbSummary()
progress.addMetric(lossMetric)
config = oa.ItTrainingConfig()
config.totalSteps = trainingSteps
config.batchSize = batchSize
config.sequenceLength = contextLength
config.sequenceUnit = "token"
config.timerName = "example_transformer_step"
training = oa.ItTraining(optimizer, config)
training.addMetric(lossMetric)
training.addCallback(progress)
training.addCallback(summary)

print("\nOA SDK Example — BPE Transformer · all-position LM")
print(f"Tokenizer: byte BPE · vocab={vocabSize} · context={contextLength}")
print(f"Model: NnTransformer(width={modelWidth}, hidden={hiddenWidth}, layers=1, heads=1)")
print(f"Params: {model.numParameters()} · AdamW(lr=0.01)")
print(f"Training: {trainingSteps} steps · batch={batchSize} · sequence={contextLength} tokens")

cursor = 0
initialLoss = 0.0
while not training.isDone():
	input, target, cursor = nextBatch(corpusTokens, cursor)
	optimizer.zeroGrad()
	with oa.GradientTape() as tape:
		logits = model.forward(input)
		loss = oa.FnLoss.crossEntropy(logits, target.reshape([target.numElements()]))
		tape.backward(loss)
	training.next(loss)
	if training.index() == 1:
		initialLoss = training.lastLoss()
training.finish()

finalLoss = training.lastLoss()
assert finalLoss < initialLoss
print("Transformer training verified: vocab=300, steps=300")
print(f"Loss: {initialLoss:.4f} -> {finalLoss:.4f}")
print(f"Generated: {generate(tokenizer)}")
# OA_DOC_END: ml-transformer

File: sdk/cpp/examples/ml/transformer.cpp
// OA_DOC_BEGIN: ml-transformer
#include <oa/oa.h>

#include <algorithm>
#include <cstdio>

namespace {
constexpr oa::I32 VocabSize = 300;
constexpr oa::I32 ContextLength = 16;
constexpr oa::I32 ModelWidth = 32;
constexpr oa::I32 HiddenWidth = 64;
constexpr oa::I32 BatchSize = 64;
constexpr oa::I32 TrainingSteps = 300;
constexpr const char* Corpus =
	"to be or not to be that is the question whether tis nobler in the mind "
	"to suffer the slings and arrows of outrageous fortune or to take arms "
	"against a sea of troubles and by opposing end them to be or not to be "
	"that is the question whether tis nobler in the mind to suffer the slings "
	"and arrows of outrageous fortune or to take arms against a sea of "
	"troubles and by opposing end them to be or not to be that is the question "
	"whether tis nobler in the mind to suffer the slings and arrows of "
	"outrageous fortune or to take arms against a sea of troubles and by "
	"opposing end them ";

oa::Matrix tokenMatrix(const oa::Vector<oa::I32>& tokens, oa::I32 batch) {
	return oa::FnMatrix::fromInt32(
		oa::Span<const oa::I32>(tokens.data(), tokens.size()),
		{batch, ContextLength}, oa::ScalarType::UInt32
	);
}

void nextBatch(const oa::Vector<oa::I32>& tokens, oa::I64& cursor,
	oa::Matrix& input, oa::Matrix& target) {
	const oa::I64 limit = static_cast<oa::I64>(tokens.size()) - ContextLength - 1;
	oa::Vector<oa::I32> x(static_cast<oa::Usize>(BatchSize * ContextLength));
	oa::Vector<oa::I32> y(x.size());
	for (oa::I32 batch = 0; batch < BatchSize; ++batch) {
		const oa::I64 start = (cursor + static_cast<oa::I64>(batch) * 7) % limit;
		for (oa::I32 position = 0; position < ContextLength; ++position) {
			const auto output = static_cast<oa::Usize>(batch * ContextLength + position);
			x[output] = tokens[static_cast<oa::Usize>(start + position)];
			y[output] = tokens[static_cast<oa::Usize>(start + position + 1)];
		}
	}
	cursor = (cursor + BatchSize) % limit;
	input = tokenMatrix(x, BatchSize);
	target = tokenMatrix(y, BatchSize);
}

oa::String generate(oa::NnTransformer& model, const oa::BpeTokenizer& tokenizer) {
	const oa::String prompt = "to be";
	auto promptTokens = tokenizer.encode(prompt.cStr());
	oa::Vector<oa::I32> context(static_cast<oa::Usize>(ContextLength), 0);
	const auto copied = std::min<oa::Usize>(promptTokens.size(), context.size());
	for (oa::Usize index = 0; index < copied; ++index) context[index] = promptTokens[index];
	oa::I32 filled = std::max<oa::I32>(1, static_cast<oa::I32>(copied));
	oa::String result = prompt;
	for (oa::I32 index = 0; index < 32; ++index) {
		auto logits = model.forward(tokenMatrix(context, 1));
		auto row = oa::FnMatrix::slice(logits, 0, filled - 1, filled);
		const auto next = static_cast<oa::I32>(oa::FnMatrix::argmax(row.reshape({VocabSize})));
		result += tokenizer.decode({next});
		if (filled < ContextLength) {
			context[static_cast<oa::Usize>(filled++)] = next;
		} else {
			for (oa::Usize token = 1; token < context.size(); ++token) context[token - 1] = context[token];
			context[ContextLength - 1] = next;
		}
	}
	return result;
}
} // namespace

OA_MAIN("ExampleMlTransformer") {
	oa::FnMatrix::setRngSeed(20260714ULL);
	oa::BpeTokenizer tokenizer(VocabSize);
	tokenizer.train(Corpus, VocabSize - 256);
	if (tokenizer.vocabSize() != VocabSize) return 1;
	const auto corpusTokens = tokenizer.encode(Corpus);

	oa::NnTransformer model(VocabSize, ContextLength, ModelWidth, HiddenWidth);
	auto parameters = model.allParameterPtrs();
	oa::AdamW optimizer(parameters, 0.01F);
	oa::MetricLoss lossMetric;
	oa::CbProgressBar progress;
	oa::CbSummary summary;
	progress.addMetric(&lossMetric);
	oa::ItTrainingConfig config;
	config.totalSteps = TrainingSteps;
	config.batchSize = BatchSize;
	config.sequenceLength = ContextLength;
	config.sequenceUnit = "token";
	config.timerName = "example_transformer_step";
	oa::ItTraining training(engine, optimizer, config);
	training.addMetric(&lossMetric);
	training.addCallback(&progress);
	training.addCallback(&summary);

	std::printf("\nOA SDK Example — BPE Transformer · all-position LM\n");
	std::printf("Tokenizer: byte BPE · vocab=%d · context=%d\n", VocabSize, ContextLength);
	std::printf(
		"Model: NnTransformer(width=%d, hidden=%d, layers=1, heads=1)\n",
		ModelWidth,
		HiddenWidth
	);
	std::printf("Params: %lld · AdamW(lr=0.01)\n", static_cast<long long>(model.numParameters()));
	std::printf(
		"Training: %d steps · batch=%d · sequence=%d tokens\n",
		TrainingSteps,
		BatchSize,
		ContextLength
	);

	oa::I64 cursor = 0;
	oa::Matrix input;
	oa::Matrix target;
	oa::F32 initialLoss = 0.0F;
	while (not training.isDone()) {
		nextBatch(corpusTokens, cursor, input, target);
		optimizer.zeroGrad();
		oa::GradientTape tape;
		auto logits = model.forward(input);
		auto loss = oa::FnLoss::crossEntropy(logits, target.reshape({target.numElements()}));
		tape.backward(loss);
		training.next(loss);
		if (training.index() == 1) {
			initialLoss = training.lastLoss();
		}
	}
	if (not training.finish().isOk()) {
		return 1;
	}

	const oa::F32 finalLoss = training.lastLoss();
	if (not (finalLoss < initialLoss)) {
		return 1;
	}
	const auto generated = generate(model, tokenizer);
	std::printf("Transformer training verified: vocab=300, steps=300\n");
	std::printf("Loss: %.4f -> %.4f\n", initialLoss, finalLoss);
	std::printf("Generated: %s\n", generated.cStr());
	return 0;
}
// OA_DOC_END: ml-transformer
